File: src/BusinessLogic/Handlers/CallbackQueryHandlers/WatchCallbackQueryHandler.cs
using Bounan.Bot.BusinessLogic.CommandDto;
using Bounan.Bot.BusinessLogic.Handlers.MessageHandlers;
using Bounan.Bot.TelegramBot.Telegram;
using Microsoft.Extensions.Logging;
using Telegram.Bot.Requests;
using Telegram.Bot.Types;

namespace Bounan.Bot.BusinessLogic.Handlers.CallbackQueryHandlers;

public class WatchCallbackQueryHandler(
    ILogger<WatchCallbackQueryHandler> logger,
    WatchMessageHandler watchMessageHandler) : ICallbackQueryHandler
{
    public static bool CanHandle(CallbackQuery item) => item.Data?.StartsWith(WatchCommandDto.Command) ?? false;

    public async Task<AnswerCallbackQueryRequest> HandleAsync(
        CallbackQuery updateItem,
        CancellationToken cancellationToken)
    {
        logger.LogInformation("Handling watch callback query {Data}", updateItem.Data);

        ArgumentNullException.ThrowIfNull(updateItem);
        ArgumentNullException.ThrowIfNull(updateItem.Data);
        ArgumentNullException.ThrowIfNull(updateItem.Message);
        ArgumentNullException.ThrowIfNull(updateItem.Message.Chat);

        await watchMessageHandler.HandleAsync(
            new Message
            {
                Chat = updateItem.Message.Chat,
                Text = updateItem.Data,
            },
            cancellationToken);

        return new AnswerCallbackQueryRequest(updateItem.Id);
    }
}

File: src/handlers/on-webhook/event-handlers/callback-query-adapters/watch-callback-query-handler.ts
import { CallbackQuery } from 'telegram-bot-api-lightweight-client';
import { CallbackQueryHandler } from '../query-handler';
import { processCallbackQuery } from './callback-query-adapter-base';
import { WatchCommandDto } from '../../command-dtos';
import { watchMessageHandler } from '../message-handlers/watch-message-handler';

const canHandle = (callbackQuery: CallbackQuery) => callbackQuery.data?.startsWith(WatchCommandDto.Command) || false;

const handler: CallbackQueryHandler = async (callbackQuery) =>
    await processCallbackQuery(callbackQuery, watchMessageHandler.handler);

export const watchCallbackQueryHandler = {
    canHandle,
    handler,
}